### PR TITLE
Reading xient becomes writer

### PR DESCRIPTION
**Pull request by eduard-bagdasaryan**

### CODE REVIEW

File: src/fs/rock/RockIoState.cc [hunk 1]
```diff
@@ -115,6 +118,13 @@ Rock::IoState::read_(char *buf, size_t len, off_t coreOff, STRCB *cb, void *data
     read.callback = cb;
     read.callback_data = cbdataReference(data);
 
+    // quit if we cannot read what they want, and the writer cannot add more
+    if (sidCurrent < 0 && writerLeft) {
```
Comment (eduard-bagdasaryan) [inline src/fs/rock/RockIoState.cc]: We can probably combine this if and the if below:

```suggestion
    if (sidCurrent < 0) {
        if (writeLeft) 
        [...]
    }
```

File: src/store/Controller.cc [hunk 2]
```diff
@@ -837,22 +809,41 @@ Store::Controller::anchorToCache(StoreEntry &entry, bool &inSync)
...
-            debugs(20, 5, "failed to anchor " << entry);
-    } else {
-        debugs(20, 7, "skipping not yet cached " << entry);
+        debugs(20, 5, "failed to anchor " << entry);
+        return true;
+    }
+
+    if (entryStatus.waitingToBeFreed) {
+        debugs(20, 5, "failed on marked unattached " << entry);
+        return true;
```
[the remark this answers is not in the document]
Comment (rousskov) [reply src/store/Controller.cc]: > since 'found' is now true when waitingToBeFreed is true.

Sorry, I do not follow. These changes do not change the values of `found` and `waitingToBeFreed` IIIRC. Also, there is no relationship between the two: `found` is about finding the matching entry in one of the stores while the `waitingToBeFreed` flag is in the transients entry itself. All four combinations of values are possible. What am I missing?

File: src/store/Controller.cc [hunk 3]
```diff
@@ -837,22 +809,41 @@ Store::Controller::anchorToCache(StoreEntry &entry, bool &inSync)
...
     }
 
-    return found;
+    if (!entryStatus.hasWriter) {
+        debugs(20, 5, "failed on abandoned-by-writer " << entry);
+        return true;
```
[the remark this answers is not in the document]
Comment (rousskov) [reply src/store/Controller.cc]: > if entryStatus.hasWriter is false the method always returns after ''if (found)`' condition.

I do not understand why you think that this method returns after `if (found)` condition if entryStatus.hasWriter is false. The two events/states (found and entryStatus.hasWriter) are independent.

Perhaps there is some confusion (on yours and/or my side) because entryStatus is calculated twice (and may change across these two methods)?

File: src/fs/rock/RockIoState.cc (hunk 1 above)
Comment (rousskov) [reply src/fs/rock/RockIoState.cc]: Agreed, but I prefer the consistency of the current "one if, one return" code structure. It might help with understanding each terminal condition better, and it reduces code changes. This is a slight preference -- both variants would be OK.

File: src/store_swapout.cc [hunk 4]
```diff
@@ -369,6 +372,14 @@ StoreEntry::mayStartSwapOut()
         return false;
     }
 
+    // in SMP mode, restrict caching to StoreEntry publisher to avoid
+    // workers releasing each other caching attempts
+    if (!Store::Root().transientsWriter(*this)) {
```
Comment (eduard-bagdasaryan) [inline src/store_swapout.cc]: Don't we need to limit the scope of this condition to SMP cases, as the comment prescribes, e.g.,
```suggestion
    if (!Store::Root()::SmpAware && !Store::Root().transientsWriter(*this)) {
```

Also, I found another problem: our unit tests (testRock and testUfs) are broken after ddc7d7e. The reason is in the following contradiction: though testRock is 'SmpAware', it does not have Transients created due to stub UsingSmp(), returning false.

Comment (rousskov) [reply src/store_swapout.cc]: > unit tests (testRock and testUfs) are broken

Glad you noticed. I probably did not run `make check`. Please fix.

Does any existing unit test create Transients, StoreMap, or any similar complex shared memory structures?

Comment (rousskov) [reply src/store_swapout.cc]: > Don't we need to limit the scope of this condition to SMP cases, as the comment prescribes

Yes, we do! The suggested change looks backwards to me though. I would

```C++
if (Store::Root()::SmpAware && !Store::Root().transientsWriter(*this)) {
```

Earlier versions checked `transientsReader()` here, but that felt too weak. I lost the (previously implied) SMP check when converting to !transientsWriter(). Root()::SmpAware() is equivalent to Controller having Transients, right?

Comment (eduard-bagdasaryan) [reply src/store_swapout.cc]: Done at 4443485.

File: src/store/Controller.cc [hunk 5]
```diff
@@ -443,6 +443,12 @@ Store::Controller::transientsWriter(const StoreEntry &e) const
     return transients && e.hasTransients() && transients->isWriter(e);
 }
 
+bool
+Store::Controller::cacheWriter(const StoreEntry &e) const
+{
+    return transients ? transientsWriter(e) : true;
```
Comment (rousskov) [inline src/store/Controller.cc]: If possible, please inline this method definition.

File: src/fs/rock/RockSwapDir.cc [hunk 6]
```diff
@@ -345,6 +349,12 @@ Rock::SwapDir::parse(int anIndex, char *aPath)
     validateOptions();
 }
 
+bool
+Rock::SwapDir::smpAware() const
+{
+    return UsingSmp();
```
Comment (rousskov) [inline src/fs/rock/RockSwapDir.cc]: Please undo this change. If this was required to fix test cases, then we need a different solution for fixing test cases.

File: src/Store.h [hunk 7]
```diff
@@ -191,10 +193,7 @@ class StoreEntry : public hash_link, public Packable
     bool hasMemStore() const { return mem_obj && mem_obj->memCache.index >= 0; }
 
     /// whether this entry can feed collapsed requests and only them
-    bool hittingRequiresCollapsing() const { return EBIT_TEST(flags, ENTRY_REQUIRES_COLLAPSING); }
-
-    /// allow or forbid collapsed requests feeding
-    void setCollapsingRequirement(const bool required);
+    bool hittingRequiresCollapsing() const { return mem_obj && (mem_obj->getReply()->pstate < Http::Message::psParsed); }
```
Comment (rousskov) [inline src/Store.h]: I know you have you checked that mem_obj is not deleted for entries that should have true hittingRequiresCollapsing(). What about entries that are created by CF initiators in non-SMP setups? Do they always get mem_obj before the first hittingRequiresCollapsing() call? If yes, can we adjust the code so that we can assert() true hittingRequiresCollapsing() just before Store::Controller::allowCollapsing() returns true?

File: src/store/Controller.cc (hunk 5 above)
Comment (rousskov) [inline src/store/Controller.cc]: Consider simplifying(?):
```suggestion
    return !transients || transientsWriter(e);
```

I do not insist on this change. Your call.

File: src/Store.h (hunk 7 above)
Comment (eduard-bagdasaryan) [reply src/Store.h]: > Do they always get mem_obj before the first hittingRequiresCollapsing() call? 

Yes, since Controller::allowCollapsing() is always called just after storeCreateEntry()/storeCreatePureEntry(), creating the memory object. I added this assert.

File: src/store/Controller.cc (hunk 5 above)
Comment (rousskov) [reply src/store/Controller.cc]: Can we solve those linking problems by adding a stub for transientsWriter()? Or is there a bigger/deeper problem here?

Comment (eduard-bagdasaryan) [reply src/store/Controller.cc]: Done, I mislead you with my comment. This inlining worked without problems I was thinking about.

File: src/clients/Client.cc [hunk 8]
```diff
@@ -506,6 +506,7 @@ void
 Client::haveParsedReplyHeaders()
 {
     Must(theFinalReply);
+
```
Comment (rousskov) [inline src/clients/Client.cc]: Please remove this out-of-scope whitespace change.

File: src/Store.h (hunk 7 above)
Comment (rousskov) [inline src/Store.h]: If code sets StoreEntry to an internally-generated (usually, but not always, as the result of an error) response, what will pstate of that response be? IIRC, for examples, look for most replaceHttpReply() callers.

File: src/store/Controller.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (rousskov) [reply src/store/Controller.cc:842]: @eduard-bagdasaryan, did my answer address your concern?

[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/store/Controller.cc:842]: > I do not understand why you think that this method returns after if (found) condition if entryStatus.hasWriter is false

IIRC, I considered the following scenario:

* this method (called from Controller::syncCollapsed() ) returns true if entryStatus.hasWriter is false, printing "failed on abandoned-by-writer" in Controller::syncCollapsed()
* a local 'found'  in Controller::syncCollapsed() becomes true, and it proceeds to "aborting unsyncable" (not to "aborting abandoned-by-writer ", which is the next 'if' block).

I am not sure whether this is a problem, i.e., whether these two states "failed on abandoned-by-writer" and "aborting unsyncable " are coherent. On the other hand, we probably should not print the two almost identical messages "failed on abandoned-by-writer" and "aborting abandoned-by-writer ".

**Comment by rousskov** [comment]
Closing this stale PR. It is a grand child of several fixes, including official PR 1117.
